### ISSUE TITLE
Issues rendering sprites and 3D models in 0.6 at the same time

### ISSUE THREAD

**Issue by cart** [bevyengine/bevy#3902]
### Discussed in https://github.com/bevyengine/bevy/discussions/3891

<div type='discussions-op-text'>

<sup>Originally posted by **cdsupina** February  7, 2022</sup>
Hello everyone! I'm trying to migrate my [game](https://github.com/thetawavegame/thetawave) from Bevy 0.5 to 0.6. The core of my game takes place in 2D, but I had a 3D background that gave my game a distinctive look. In Bevy 0.5 I was using a perspective camera and didn't have any issues rendering the 2D sprites and the background 3D models. Below is the Bevy 0.5 code for the perspective camera I was using.

```rust
// Setup perspective camera in Bevy 0.5
commands.spawn_bundle(PerspectiveCameraBundle {
    transform: Transform::from_xyz(0.0, 0.0, 950.0)
        .looking_at(Vec3::ZERO, Vec3::Y),
    perspective_projection: PerspectiveProjection {
        far: 10000.0,
        ..Default::default()
    },
    ..Default::default()
});
```

Snapshot of game from Bevy 0.5:
![Capture](https://user-images.githubusercontent.com/15306815/152926560-b2cd6382-00f2-400c-a5d9-375f67d99625.PNG)

Since migrating to 0.6. I haven't been able to setup the camera in a way that lets me view the background 3D models and the sprites. Below is the perspective camera code I was using in Bevy 0.6 after I initially migrated (same as before).

```rust
// Setup perspective camera in Bevy 0.6
commands.spawn_bundle(PerspectiveCameraBundle {
    transform: Transform::from_xyz(0.0, 0.0, 950.0)
        .looking_at(Vec3::ZERO, Vec3::Y),
    perspective_projection: PerspectiveProjection {
        far: 10000.0,
        ..Default::default()
    },
    ..Default::default()
});
```

The image below is the game in Bevy 0.6. 3D models are visible but sprites are not appearing.

![Capture](https://user-images.githubusercontent.com/15306815/152923385-74a99baf-a4c5-4ff8-96ec-6e0e30d0c0ac.PNG)

The next thing I tried was a 2D orthographic camera using the following code:

```rust
let mut camera = OrthographicCameraBundle::new_2d();
camera.transform = Transform::from_xyz(0.0, 0.0, 1000.0);
camera.orthographic_projection.far = 10000.0;
```

Now the 2D sprites are appearing but not the background models.
![Capture](https://user-images.githubusercontent.com/15306815/152926358-1d9139a5-a42f-40d0-a1dc-fee44b754ff2.PNG)

I was wondering if anyone had insight into how to render both the sprites and the background models like I had from Bevy 0.5 with the perspective camera. I also tried playing around with the other camera properties but didn't have much luck. Thank you!
</div>

**Comment by cart** [comment]
This is probably partially because 3d renders _after_ 2d in 0.6:

https://github.com/bevyengine/bevy/blob/main/crates/bevy_core_pipeline/src/main_pass_driver.rs

However switching the order _does not_ resolve the issue:

```rust
impl Node for MainPassDriverNode {
    fn run(
        &self,
        graph: &mut RenderGraphContext,
        _render_context: &mut RenderContext,
        world: &World,
    ) -> Result<(), NodeRunError> {
        let extracted_cameras = world.get_resource::<ExtractedCameraNames>().unwrap();
        
        if let Some(camera_3d) = extracted_cameras.entities.get(CameraPlugin::CAMERA_3D) {
            graph.run_sub_graph(
                crate::draw_3d_graph::NAME,
                vec![SlotValue::Entity(*camera_3d)],
            )?;
        }

        if let Some(camera_2d) = extracted_cameras.entities.get(CameraPlugin::CAMERA_2D) {
            graph.run_sub_graph(
                crate::draw_2d_graph::NAME,
                vec![SlotValue::Entity(*camera_2d)],
            )?;
        }
        Ok(())
    }
}
```

We should definitely do that, as 2d-on-top-of-3d is a way more common scenario.  But this will require some investigation to resolve.

**Comment by Frank-Buss** [comment]
Any solution to this problem? Looks like I have the same problem with bevy_prototype_lyon. But what I don't understand: UiCameraBundle and buttons work on top of 3D. Is this a different camera?

**Comment by djee-ms** [comment]
I would propose to use layer-based rendering. This would solve not only this, but also the problem of having multiple cameras. Currently the "active" camera is selected for each mode (2d / 3d / UI) based on its `Name` (ugh?!) and stored in `ActiveCameras`. This is unexpected and doesn't allow having multiple cameras at the same time, and if you do one randomly gets picked up. This also means renaming a camera breaks it.

The "layered" approach is just what Unity does: each camera renders the object in a given layer or set of layers. All cameras are composited together into the final render target. We just need an order for the layers themselves, which is the order in which the final blending happens. We can have a "primary" camera to bypass and render directly into the swap chain if needed. With this, 2D and 3D can render into separate framebuffers and we composite them at the end of the frame, in the order in which the layers of each camera is defined.

**Comment by djee-ms** [comment]
> Any solution to this problem? Looks like I have the same problem with bevy_prototype_lyon. But what I don't understand: UiCameraBundle and buttons work on top of 3D. Is this a different camera?

Yes that's a third, different camera. And that one composites correctly with 3D.

**Comment by Frank-Buss** [comment]
> The "layered" approach is just what Unity does: each camera renders the object in a given layer or set of layers. All cameras are composited together into the final render target. We just need an order for the layers themselves, which is the order in which the final blending happens. We can have a "primary" camera to bypass and render directly into the swap chain if needed. With this, 2D and 3D can render into separate framebuffers and we composite them at the end of the frame, in the order in which the layers of each camera is defined.

Sounds good. Can this be done with the current Bevy 0.6 release? And do you have a simple example how this code:
https://github.com/Nilirad/bevy_prototype_lyon/blob/master/examples/readme.rs
would also show a 3D cube?

**Comment by djee-ms** [comment]
That's a large change and complex change, because it touches the internals of the renderer and I'm not familiar enough with all its parts. This would need coordination with the rendering folks and @cart to approve the design beforehand. I doubt you can do that change purely from app code, since you need to modify the render graph at least.

**Comment by djeedai** [comment]
Also posted with the wrong account (meh 😅); please consider all my above posts personal opinions, unrelated in any way to my employer (Microsoft).

**Comment by Frank-Buss** [comment]
Ok, not really related to this problem, but I posted it in the bevy_prototype_lyon project, maybe they know how to solve at least my problem, and maybe it helps with the original problem here, too.
https://github.com/Nilirad/bevy_prototype_lyon/issues/145

**Comment by cart** [comment]
My current plan for RenderTargets should accommodate this:

* A RenderTarget entity can draw to any supported target (window or arbitrary texture)
* Cameras can assign themselves to a RenderTarget (and default to the primary window)
* Cameras decide what layers to draw, entities can assign themselves to layers (this is already possible)
* RenderTargets default to using the "main" render graph, so if a sprite entity assigns itself to a layer on a render target, by default it would render like a normal sprite would
* RenderTargets can define their draw order relative to other render targets (this api will need some design work, at first im thinking we'll just handle all render targets prior to the main pass ... just to get the ball rolling)

With this, you could composite the results of arbitrary render targets (with arbitrary entities) in arbitrary orders in the default "primary window" target. If you use something like sprites or ui to render the targets relative to each other, you could do this all in the "main app" without writing any custom render logic.

**Comment by cart** [comment]
Some initial render-to-texture work has been merged here: #3377, but it requires manual render graph interventions to make it work / doesnt have the entity driven UX yet.

**Comment by cart** [comment]
That being said: we still need to actually fix the default "2d on top of 3d" behavior. People shouldn't need render targets for that.

**Comment by agg23** [comment]
Are there any planed 2D mixed with 3D changes coming to 0.7?

**Comment by geckoxx** [comment]
> 

The changed render order fixes it but only without MSAA. Because with MSAA the 2d and 3d render passes uses there own sampled_targets which then overrides the complete view. The ui render pass does not use MSAA so it is not affected by this issue.

With changed render order and with msaa:
![with_msaa](https://user-images.githubusercontent.com/688816/162611866-adeb4930-104c-4f7d-974c-ce72c5d3be79.png)
With changed render order and without msaa:
![without_msaa](https://user-images.githubusercontent.com/688816/162611871-0d23d93a-7d1f-423e-afa9-14db70e58889.png)

**Comment by HackerFoo** [comment]
#3552 should fix this. Also, #3568 allows controlling when depth is cleared to allow layering multiple cameras.

**Comment by geckoxx** [comment]
I tested #3552 with the changed render order and it fixes this issue.

**Comment by cart** [comment]
This should be resolved by #4467
